dusk/grammar.py

```diff
@@ -144,24 +145,59 @@ def field_declaration(self, name: str, type: expr):
         self.add_field_declaration(name, type, False)
 
     def add_field_declaration(self, name: str, type: expr, is_temporary: bool):
-        self.ctx.scope.current_scope.add(
-            name, DuskField(make_field(name, self.field_type(type), is_temporary))
-        )
+        h_dim = self.field_type(type)
+        if h_dim is not None:
+            self.ctx.scope.current_scope.add(
+                name, DuskField(make_field(name, h_dim, is_temporary)),
+            )
+        else:
+            self.ctx.scope.current_scope.add(
+                name, DuskField(make_vertical_field(name, is_temporary)),
+            )
 
     def type(self, node):
         return dispatch({Subscript: self.field_type}, node)
 
     @transform(
         Subscript(
             value=name("Field"),
-            slice=Index(value=Capture(_).to("location_chain")),
+            slice=Index(
+                value=OneOf(
+                    Tuple(
+                        elts=FixedList(
+                            Capture(OneOf(Compare, Name)).to("hindex"),
+                            Capture(Name).to("vindex"),
+                        ),
+                        ctx=Load,
+                    ),
+                    # FIXME: ensure built-ins (like `Edge`) aren't _shadowed_ by variables
+                    # TODO: hardcoded string
+                    Capture(OneOf(Compare, name(OneOf("Edge", "Cell", "Vertex")))).to(
+                        "hindex"
+                    ),
```

Comment:
Just a note: we could consider some infrastructure to support _overloading_ or _optional positional arguments by type_ properly. This is the second time we encounter this (we also have this for field indexing). I feel like _overloading_ makes the code quite readable and intuitive, so we might use something like this more often in the future too.